servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -441,54 +462,125 @@ public Completable closeAsyncGracefully() {
+            this.healthCheckInterval = healthCheckInterval;
+            this.failedThreshold = failedThreshold;
+        }
+    }
+
+    private static final class Host<Addr, C extends LoadBalancedConnection> implements ListenableAsyncCloseable {
+        private static final Object STATE_ACTIVE = new Object();
+        private static final Object STATE_EXPIRED = new Object();
+        private static final Object STATE_CLOSED = new Object();
+
+        private static final ConnState EMPTY_CONN_STATE = new ConnState(EMPTY_ARRAY, STATE_ACTIVE);
+        private static final ConnState CLOSED_CONN_STATE = new ConnState(EMPTY_ARRAY, STATE_CLOSED);
 
         @SuppressWarnings("rawtypes")
         private static final AtomicReferenceFieldUpdater<Host, ConnState> connStateUpdater =
                 AtomicReferenceFieldUpdater.newUpdater(Host.class, ConnState.class, "connState");
 
-        private enum State {
-            ACTIVE,
-            EXPIRED,
-            CLOSED
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<Host> connectionFailuresUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(Host.class, "connectionFailures");
+
+        private static final class HealthCheck<ResolvedAddress, C extends LoadBalancedConnection>
+                implements Runnable, Cancellable {
+            private final Predicate<C> selector;
+            private final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory;
+            private final Host<ResolvedAddress, C> host;
+            private volatile boolean cancelled;
+
+            private HealthCheck(final Predicate<C> selector,
+                               final ConnectionFactory<ResolvedAddress, ? extends C> connectionFactory,
+                               final Host<ResolvedAddress, C> host) {
+                this.selector = selector;
+                this.connectionFactory = connectionFactory;
+                this.host = host;
+            }
+
+            public void schedule() {
+                host.healthCheckConfig.executor.schedule(this, host.healthCheckConfig.healthCheckInterval);
+            }
+
+            @Override
+            public void run() {
+                if (cancelled) {
+                    return;
+                }
+                connectionFactory.newConnection(host.address, null)
+                        .flatMapCompletable(newCnx -> {
+                            if (!selector.test(newCnx)) {
+                                return newCnx.closeAsync().concat(Completable.failed(new RuntimeException()));
+                            }
+                            if (host.addConnection(newCnx)) {
+                                host.markHealthy();
+                            } else {
+                                return newCnx.closeAsync().concat(Completable.failed(new RuntimeException()));
+                            }
+                            return completed();
```

Comment:
Consider aadding more debug logs when:
1. heath-check is sceduled
2. result of each health-check attempt
3. when health-check is cancelled